File: ruby/roster.rb
require_relative 'student'

class Roster
  attr_reader :class_roster_array

  EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i.freeze

  def initialize
    @class_roster_array = []
  end

  def add(student_info)
    student_info_array = student_info.split(',')
    @class_roster_array << Student.new(*student_info_array[0..4], student_info_array[5..7], student_info_array[8])
  end

  def remove(student_id)
    student = find_student(student_id)
    if student
      @class_roster_array.delete(student)
    else
      puts 'A student with this ID was not found.'
    end
  end

  def print_all
    class_roster_array.each(&:print)
  end

  def print_average_days_per_course(student_id)
    student = find_student(student_id)
    puts student.days_per_course.sum / 3
  end

  def print_invalid_emails
    @class_roster_array.each do |student|
      puts student.email unless student.email =~ EMAIL_REGEX
    end
  end

  def print_by_major(major)
    @class_roster_array.each { |student| student.print if student.major == major }
  end

  private

  def find_student(student_id)
    @class_roster_array.find { |student| student.student_id == student_id }
  end
end

File: c++/roster.cpp
#include "roster.h"
#include "student.cpp"
#include <iostream>

using namespace roster;
using namespace student;
using namespace std;

Roster::Roster(){ m_nbrStudents = 0; };

Roster::~Roster() { 
  for (int i = 0; i < m_nbrStudents; i++) {
    delete classRosterArray[i];
  }
};

void Roster::add(string studentID, string firstName, string lastName,
                 string emailAddress, int age, int daysInCourse1,
                 int daysInCourse2, int daysInCourse3,
                 degree::DegreeProgram degreeProgram) {
  Student *newStudent =
      new Student(studentID, firstName, lastName, emailAddress, age, daysInCourse1,
              daysInCourse2, daysInCourse3, degreeProgram);
  classRosterArray[m_nbrStudents] = newStudent;
  m_nbrStudents++;
}

void Roster::remove(string studentID) {
  for (int i = 0; i < m_nbrStudents; i++) {
    if (getStudent(i).getStudentID() == studentID) {
      delete classRosterArray[i];
      if (i < m_nbrStudents - 1) {
        classRosterArray[i] = classRosterArray[m_nbrStudents - 1];
      }
      m_nbrStudents--;
      return;
    }
  }
  cout << "A student with this ID was not found." << endl;
}

void Roster::printAll() {
  for (int i = 0; i < m_nbrStudents; i++) {
    getStudent(i).print();
  }
}

void Roster::printAverageDaysInCourse(string studentID) {
  for (int i = 0; i < m_nbrStudents; i++) {
    if (getStudent(i).getStudentID() == studentID) {
      int *daysInCourses = getStudent(i).getDaysInCourses();
      int averageDaysInCourse =
          (daysInCourses[0] + daysInCourses[1] + daysInCourses[2]) / 3;
      cout << "Average days in course for student #" << studentID << ": "
           << averageDaysInCourse << endl;
      return;
    }
  }
  cout << "A student with this ID was not found." << endl;
}

void Roster::printInvalidEmails() {
  for (int i = 0; i < m_nbrStudents; i++) {
    string emailAddress = getStudent(i).getEmailAddress();
    if (emailAddress.find('@') == string::npos ||
        emailAddress.find('.') == string::npos ||
        emailAddress.find(' ') != string::npos) {
      cout << emailAddress << endl;
    }
  }
}

void Roster::printByDegreeProgram(degree::DegreeProgram degreeProgram) {
  for (int i = 0; i < m_nbrStudents; i++) {
    if (getStudent(i).getDegreeProgram() == degreeProgram) {
      getStudent(i).print();
    }
  }
}

Student Roster::getStudent(int idx) { return *classRosterArray[idx]; }

int Roster::getNbrStudents() { return m_nbrStudents; }
